### ISSUE TITLE
How do I drag nodes around like in most d3 examples?

### ISSUE THREAD

**Issue by aprotopapas** [dagrejs/dagre-d3#74]
Hi Chris,

I have managed to disable zooming and dragging (of the whole graph), but I am struggling to find instructions on how to make the graph more interactive, i.e. be able to drag nodes around while maintaining the graph structure.

Thanks
Andreas

**Comment by AhmedRiyad** [comment]
@ion999 Yes, it is possible by using manhattan option in jointjs.

```javascript
function makeLink(parentElementLabel, childElementLabel) {

                    return new joint.dia.Link({
                        source: {id: parentElementLabel},
                        target: {id: childElementLabel},
                        router: {name: 'manhattan'},
                        connector: {name: 'normal'},
                        attrs: {'.marker-target': {d: 'M 4 0 L 0 2 L 4 4 z'}}

                    });
                }
```

**Comment by ion999** [comment]
manhattan option looks good, but can i draw the links without visually overlapping? (and still keep manhattan attributes)

**Comment by cueqzapper** [comment]
@cpettitt I managed to create a dagre d3 which allows to zoom and pan. It also allows draging for nodes. I followed your suggestions and used the most recent version of dagre d3. 

My problem is, that I could not complete the second part of your suggestion. How do I have to update the edges?

Here you can find a minimal setup: http://codepen.io/anon/pen/XJZrxm

**Comment by jennyxing27** [comment]
@cueqzapper I got dragging for both nodes and edges working. All the code is at the very bottom. like @cpettitt said, I copied the calcPoints and intersectRect functions from the dagre-d3 source code and modified them a bit to fit my code. 

The code is here: http://jsfiddle.net/z41bx2cu/

lemme know if you have any questions.

**Comment by cueqzapper** [comment]
Thx jennyxing. We decided to use graphviz to create our graphs, because dagre-d3 is sometimes not as perfect as we needed it on very large and complex graphs. But your example works very nice! Well done and thanks for sharing!

**Comment by davidheryanto** [comment]
Hi @jennyxing27 Thanks for the jsfiddle code, it's really helpful. Do you happen to know how to make the edgeLabel moves as well when you drag the nodes/edges? I've been trying to figure this out but still no luck.

**Comment by saimaparveen** [comment]
@jennyxing27 The jsfiddle you provided is not working in IE.
I am able to drag nodes in IE but edges are not getting dragged.

**Comment by jennyxing27** [comment]
@saimaparveen what version IE are you using? I can try to look into it.

**Comment by motoz** [comment]
I managed to get the edge labels to move http://jsfiddle.net/egfx43hs/4/ (sort of, the position is not recalculated, just moved along with the node being dragged)

Thanks @jennyxing27 for the jsfiddle, just what I needed!

**Comment by saimaparveen** [comment]
@jennyxing27  I tried with IE11.
It is showing error in line number 220.
Unable to get property 1 of undefined.

![screenshot 303](https://cloud.githubusercontent.com/assets/9054196/9277444/b9d42dda-42c7-11e5-8f39-42c22088dbf4.png)

**Comment by motoz** [comment]
@saimaparveen, IE apparently leaves out the zero completely from the transform attribute when the 'y' part is zero (ie. transform(10 0) in firefox is transform(10) in IE), which the regex didn't handle. This one should work: http://jsfiddle.net/egfx43hs/11/

**Comment by saimaparveen** [comment]
@motoz , the fiddle that you gave is not working... I am unable to drag node in IE..
It is not giving any errors

**Comment by motoz** [comment]
@saimaparveen, I know it does not work in IE. The IE error you posted above "Unable to get property 1 of undefined" was from my addition to @jennyxing27 code, that also moves the edge-labels in addition to nodes and edges, so I checked why IE gave an error on that and fixed it. I have no idea why IE don't work with @jennyxing27 code.

**Comment by jennyxing27** [comment]
@motoz sorry about my code. I was in a hurry and didn't do cross browser testing. thanks for making it work in IE!

**Comment by motoz** [comment]
@jennyxing27, sorry but I didn't fix your code for IE, I only fixed my own addition that  @saimaparveen said gave the error "Unable to get property 1 of undefined" in IE. Your code works perfectly for me on firefox and chrome, thanks again, but as far as I understand the edges does not move when dragging nodes in IE and I have not looked at that at all.

**Comment by djangosdk** [comment]
@cpettitt
Can you please update demo examples with motoz`fiddle. Can be helpful for others.

**Comment by djangosdk** [comment]
@cpettitt , @motoz 
Please see attached screenshot. When I try to move left most block of elements to far right by dragging,
Edges are ot straight anymore as they were initially.
![problem](https://cloud.githubusercontent.com/assets/4101480/15104508/d9639580-15d2-11e6-939f-67d47858f22d.png)

**Comment by motoz** [comment]
Yes, but you can drag the edges back into a straight line by carefully positioning the mouse anywhere on the line.

**Comment by djangosdk** [comment]
@motoz:
You are right. Previously I was dragging edges points, didnt know I can also drag curve length to adjust it. 
But Should not edges be calculated again for the node while dragging node ? Shortest dance to follow and so on ? I mean programmatically ?

**Comment by motoz** [comment]
Of course, that would be great.

**Comment by chruetli** [comment]
Is there any idea to make clusters (compound:true) work?

**Comment by sharmisthab04** [comment]
Can you please update demo examples for IE workaround for drag node and edges ...it will be very helpful

**Comment by graphdude** [comment]
@motoz
can you please help me with drag behaviour here?
Following your same code in angular 4 with d3 version 4
I ma getting following error on var nodedrag..it's undefined somehow.
I am stuck.


![img_20180809_184610](https://user-images.githubusercontent.com/41581899/44741212-0e2b0580-ab1a-11e8-9e4c-ffc101923021.jpg)

**Comment by motoz** [comment]
Sorry, I have no idea. I haven't looked at this since I used it in a small project three years ago and hacked on it a couple of hours to get the labels to move when using the addition made by @jennyxing27 to enable dragging.

**Comment by graphdude** [comment]
@motoz
Okay.. Thanks for the reply though.
Any idea how to add navigation buttons?
How to change x,y on button click?
The way pan is possible with mouse..

**Comment by Luttik** [comment]
@motoz @mafar Fixing the edge behaviour is quite an easy fix. It occurs because if you translate the start and end node of an edge the edge itsself is translated twice. Therefore translate it with half the distance of the edges translation and you fix the odd behaviour.

http://jsfiddle.net/luttik/fnrh53b1/3/

**Comment by tushevskyi** [comment]
@Luttik hello, with the last update of dagre or dagre-d3 drag and drop start working not so perfect as on your jsfiddle. The arrows are missing the coordinates of the node and move to the node center after the start of the drag. Do you have any idea how it can be solved?

**Comment by Luttik** [comment]
@tushevskyi I don't know this sounds like a separate issue. I have only ever considered to use dagre-3d but never actually came around to it so I am probably the wrong person to ask.

**Comment by tushevskyi** [comment]
@Luttik thanks anyway)

**Comment by rkma** [comment]
How to set restriction for Fixing dragging position in outside of bounding area?

**Comment by yashinigam** [comment]
Hi,

Has someone tried dagre drag and drop on d3 V4. The drag and drop doesnt seem to work as mentioned here - http://jsfiddle.net/bbxvhx17/4/

The nodes move fine, but not the edges. 

Any help is much appreciated.
